### ISSUE TITLE
Reflexões [3]

### ISSUE THREAD

**Issue by tacsio** [gustavoprimolan/jornada-dev-eficiente#3]
_Aqui tem uma dúvida que até marquei para debater em alguma das lives._

Concorda que aqui é possível instanciar um Pedido inválido (com o valor total maior ou menor que o valor dos itens em si)?

https://github.com/gustavoprimolan/jornada-dev-eficiente/blob/f13b258b717450f31390f1ef249372c69759de7a/casadocodigo/src/main/java/br/com/casadocodigo/forms/PurchaseForm.java#L75-L80

**Comment by gustavoprimolan** [comment]
Realmente, @tacsio . Acho que a validação desse valor realmente deveria ser na construção do objeto. O que acho que talvez a validação dele não seria necessário, né? Já que já possuo os valores no banco de dados e poderia calcular o valor total só no backend, entretanto como está descrito na atividade que o client está enviando isso, então acho que vamos ter que validar mesmo.

Vou fazer essa alteração ai você da uma olhada de novo. Valeu pela issue, cara!

**Comment by tacsio** [comment]
Pensei na mesma coisa logo no início dessa implementação.

Um ponto interessante é que essa abordagem (de sempre calcular pelo backend), no tempo, só seria válida:
1. Se os dados dos itens (quantidade e valor) fossem salvos (não apenas o relacionamento com o livro)

Ou 

2. Se o valor do livro fosse imutável


Assim o valor total de um pedido teria a integridade necessária.

Mas isso é um ponto interessante de debate. 

Numa eventual funcionalidade de "editar pedido" (por exemplo excluir um item do pedido que está no estado iniciado) o valor dos itens já presentes num pedido não deveria mudar (vai que comprou numa promoção, mas desistiu de algum item), se usarmos o valor presente na entidade do livro, esse valor pode ser diferente impedindo essa validação.

Não se se foi viagem demais esse ponto (pensei em implementar mas era muito distante da espec do probelma).

**Comment by gustavoprimolan** [comment]
Muita boa sua análise, @tacsio . Nem me passou pela cabeça sobre o valor dos livros serem mutáveis, mas é algo que realmente acontece em um ambiente real. Os valores das coisas são mutáveis o tempo todo. Acho que faz sentido eu criar mais um campo para o item e guardar o valor do livro naquela hora em que ele foi comprado.
Não sei se estou enganado, mas normalmente, em um ambiente real, os pedidos costumam ser imutáveis, apenas possibilitando o cancelamento ou a criação de um novo, mas realmente vendo por esse ponto do item em promoção e querer cancelar algum pedido é algo interessante para se pensar tbm.

Novamente, cara. Muito legal sua análise. Acho que devemos nos aproximar o máximo de uma aplicação real e isso é algo que eu ainda preciso ser mais crítico.

**Comment by tacsio** [comment]
Exatamente, @gustavoprimolan , também acredito que em ambiente real devem ser imutáveis.

Talvez até podemos puxar esse ponto em alguma das lives ou mesmo no discord. Mesmo com o objetivo bem claro no texto, existem cenários que são bem interessantes de extrapolar...  Se a ideia é "treinar" com a prática deliberada, esse tipo de extrapolação seria muito legal para o problema.

_Tem até uma loucura que geralmente faço quando me cadastro em e-commerces (não tem muito haver com imutabilidade de pedido, mas sim segurança), que é ao fazer o cadastro usar uma senha "tosca"... e assim que criar, pedir para recuperar a senha.... cara a quantidade de lugar (uns até grandes) que o email de recuperação é a própria senha plain text é absurda kkkkkk_

**Comment by asouza** [comment]
Extrapolem!!!! Pelo amor de deus! Essa é uma possibilidade muito legal de treinar em ambiente propício! E a discussão é super válida! Quando a gera uma compra realmente precisamos desconectar o valor final das coisas da referência original, justamente pq não queremos sofrer efeito colateral de uma edição.

**Comment by tacsio** [comment]
Olha aí @gustavoprimolan .... feature liberada pelo @asouza 
ehehhehe
